hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/BucketIndexSupport.scala

```diff
@@ -0,0 +1,164 @@
+ */
+
+package org.apache.hudi
+
+import org.apache.hadoop.fs.FileStatus
+import org.apache.hudi.common.config.HoodieMetadataConfig
+import org.apache.hudi.common.fs.FSUtils
+import org.apache.hudi.common.table.HoodieTableConfig
+import org.apache.hudi.config.HoodieIndexConfig
+import org.apache.hudi.index.HoodieIndex
+import org.apache.hudi.index.HoodieIndex.IndexType
+import org.apache.hudi.index.bucket.BucketIdentifier
+import org.apache.log4j.LogManager
+import org.apache.spark.sql.catalyst.expressions
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, EmptyRow, Expression, Literal}
+import org.apache.spark.sql.types.{DoubleType, FloatType}
+import org.apache.spark.util.collection.BitSet
+
+import scala.collection.{JavaConverters, mutable}
+
+class BucketIndexSupport(metadataConfig: HoodieMetadataConfig) {
+
+  private val log = LogManager.getLogger(getClass);
+
+  /**
+   * Returns the configured bucket field for the table
+   */
+  private def getBucketHashField: Option[String] = {
+    val bucketHashFields = metadataConfig.getString(HoodieIndexConfig.BUCKET_INDEX_HASH_FIELD)
+    if (bucketHashFields == null) {
+      val recordKeys = metadataConfig.getString(HoodieTableConfig.RECORDKEY_FIELDS)
+      if (recordKeys == null) {
+        Option.apply(null)
+      } else {
+        val recordKeyArray = recordKeys.split(",")
+        if (recordKeyArray.length == 1) {
+          Option.apply(recordKeyArray(0))
+        } else {
+          log.warn("bucket query index only support one bucket field")
+          Option.apply(null)
+        }
+      }
+    } else {
+      val fields = bucketHashFields.split(",")
+      if (fields.length == 1) {
+        Option.apply(fields(0))
+      } else {
+        log.warn("bucket query index only support one bucket field")
+        Option.apply(null)
+      }
+    }
+  }
+
+  def getCandidateFiles(allFiles: Seq[FileStatus], bucketIds: BitSet): Set[String] = {
+    val candidateFiles: mutable.Set[String] = mutable.Set.empty
+    for (file <- allFiles) {
+      val fileId = FSUtils.getFileIdFromFilePath(file.getPath)
+      val fileBucketId = BucketIdentifier.bucketIdFromFileId(fileId)
+      if (bucketIds.get(fileBucketId)) {
+        candidateFiles += file.getPath.getName
+      }
+    }
+    candidateFiles.toSet
+  }
+
+  def filterQueriesWithBucketHashField(queryFilters: Seq[Expression]): Option[BitSet] = {
+    val bucketNumber = metadataConfig.getInt(HoodieIndexConfig.BUCKET_INDEX_NUM_BUCKETS)
+    val bucketHashFieldOpt = getBucketHashField
+    if (bucketHashFieldOpt.isEmpty || queryFilters.isEmpty) {
+      None
+    } else {
+      val matchedBuckets = getExpressionBuckets(queryFilters.reduce(And), bucketHashFieldOpt.get, bucketNumber)
+
+      val numBucketsSelected = matchedBuckets.cardinality()
+
+      // None means all the buckets need to be scanned
+      if (numBucketsSelected == bucketNumber) {
+        log.info("bucket query match all file slice, fallback other index")
```

Comment:
Do you mean all the hash fields should be included?

Reply:
all the buckeId is matched, so fallback to other index.
for example, I bucket field with `a` bucket number, condition expression is `a = 1 Or b = 'abc'`, all bucketId will be include, it mean we need scan all file slice, so fallback use other index to try.

Reply:
file slice is an internal notion, maybe you should say `The query predicates does not specify equality for all the hasing fields, ...`

Reply:
sounds good